### PR TITLE
feat(util-endpoint): add endpoint ruleset cache

### PR DESCRIPTION
**Pull request by kuhe**
https://github.com/aws/aws-sdk-js-v3/issues/6423

This PR creates a cache for endpoint ruleset resolution.

As one would expect, the input and cache key are the endpoint params and the output and cache value is the endpoint.

Some services declare endpoint parameters in excess of what is actually evaluated in the ruleset, so codegen provides the explicit pared param list to the cache instance.

- [x] codegen to connect EndpointCache to the service's endpointResolver.ts file contents.
- [x] codegen to determine what the pared parameter list is to avoid cache pollution from irrelevant params

### CODE REVIEW

**Comment by kuhe** [comment]
The S3 resolver would now look like this below.
Notably absent is `Key` from the parameter list.

```ts
// smithy-typescript generated code
import { awsEndpointFunctions } from "@aws-sdk/util-endpoints";
import { EndpointV2, Logger } from "@smithy/types";
import { customEndpointFunctions, EndpointCache, EndpointParams, resolveEndpoint } from "@smithy/util-endpoints";

import { EndpointParameters } from "./EndpointParameters";
import { ruleSet } from "./ruleset";

const cache = new EndpointCache({
  size: 50,
  params: [
    "Accelerate",
    "Bucket",
    "DisableAccessPoints",
    "DisableMultiRegionAccessPoints",
    "DisableS3ExpressSessionAuth",
    "Endpoint",
    "ForcePathStyle",
    "Region",
    "UseArnRegion",
    "UseDualStack",
    "UseFIPS",
    "UseGlobalEndpoint",
    "UseObjectLambdaEndpoint",
    "UseS3ExpressControlEndpoint",
  ],
});

export const defaultEndpointResolver = (
  endpointParams: EndpointParameters,
  context: { logger?: Logger } = {}
): EndpointV2 => {
  return cache.get(endpointParams as EndpointParams, () =>
    resolveEndpoint(ruleSet, {
      endpointParams: endpointParams as EndpointParams,
      logger: context.logger,
    })
  );
};

customEndpointFunctions.aws = awsEndpointFunctions;
```

**Comment by kuhe** [comment]
In one benchmark, a batch of 10 parallel Amazon S3 HeadObject requests completed in 80ms, as opposed to 86ms without this cache.

File: packages/util-endpoints/src/cache/EndpointCache.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+  private parameters: string[] = [];
+
+  /**
+   * @param [params] - list of params to consider as part of the cache key.
+   *
+   * If the params list is not populated, all object keys will be considered.
+   * This may be out of order depending on how the object is created and arrives to this class.
+   */
+  public constructor({ size, params }: { size?: number; params?: string[] }) {
+    this.capacity = size ?? 50;
```
Comment (trivikr) [inline packages/util-endpoints/src/cache/EndpointCache.ts:20]: nit: This is already initialized as `50`on line 9, and doesn't have to be overridden
```suggestion
    if (size) {
      this.capacity = size;
    }
```

**Review by trivikr — APPROVED** [review]

File: smithy-typescript-codegen/src/test/java/software/amazon/smithy/typescript/codegen/endpointsV2/RuleSetParameterFinderTest.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+                                },
+                                true
+                              ]
+                            },
+                            true
+                          ]
+                        }
+                      ],
+                      "endpoint": {
+                        "url": "https://www.{BasicParameter}.{UrlOnlyParameter}.com",
```
Comment (milesziemer) [inline smithy-typescript-codegen/src/test/java/software/amazon/smithy/typescript/codegen/endpointsV2/RuleSetParameterFinderTest.java:89]: Could you include a test with a more complex url, maybe with a [template shorthand](https://smithy.io/2.0/additional-specs/rules-engine/specification.html#template-shorthand), as a sanity check on the regex and to make sure we don't break it down the road?

File: packages/util-endpoints/src/cache/EndpointCache.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+  public size() {
+    return this.data.size;
+  }
+
+  private hash(endpointParams: EndpointParams): string {
+    let buffer = "";
+    const params = this.parameters.length ? this.parameters : Object.keys(endpointParams);
+    for (const param of params) {
+      buffer += endpointParams[param] ?? "" + "|";
```
Comment (milesziemer) [inline packages/util-endpoints/src/cache/EndpointCache.ts:58]: Some questions about this hashing method:
1. Does the specific `param` not matter for the hash? At least in the case where `params` is going to be `Object.keys(endpointParams)`, seems like it could matter
2. Can `endpointParams[param]` contain a `|` that would mess with the hash? 
3. Should anything be appended in the case where the key `param` doesn't exist? Might matter in the same case as 1.

Maybe these are edge cases that don't sufficiently impact the effectiveness of the hash though.

File: smithy-typescript-codegen/src/test/java/software/amazon/smithy/typescript/codegen/endpointsV2/RuleSetParameterFinderTest.java (hunk 2 above)
Comment (kuhe) [reply smithy-typescript-codegen/src/test/java/software/amazon/smithy/typescript/codegen/endpointsV2/RuleSetParameterFinderTest.java:89]: added test case for shorthand, which I didn't know existed

File: packages/util-endpoints/src/cache/EndpointCache.ts (hunk 3 above)
Comment (kuhe) [reply packages/util-endpoints/src/cache/EndpointCache.ts]: 1. params needs to be consistently sorted, so parameters are alphabetized by codegen. In the case where Object.keys are used, this is best-effort since they are not guaranteed to be in any order.
2. a parameter with a delimiter in it can affect the hash, yes. `A="|",B=void` and `A=void,B="|"` would both hash to `|||`. I'm going to make the delimiter `|;` and abandon caching if this sequence exists in a value.
3. A delimiter as placeholder is enough in this case.

File: smithy-typescript-codegen/src/test/java/software/amazon/smithy/typescript/codegen/endpointsV2/RuleSetParameterFinderTest.java (hunk 2 above)
Comment (milesziemer) [reply smithy-typescript-codegen/src/test/java/software/amazon/smithy/typescript/codegen/endpointsV2/RuleSetParameterFinderTest.java:101]: I mean one so that it has the `#` in the url, sorry. Just so we make sure those kinds of urls are matched by the regex.

Comment (milesziemer) [reply smithy-typescript-codegen/src/test/java/software/amazon/smithy/typescript/codegen/endpointsV2/RuleSetParameterFinderTest.java:101]: Nevermind, practically speaking we're only looking for endpoint param names, so the object path won't matter.
